File: structy/02-linked-list/06-reverse-linked-list/01.js
/*
reverse list
Write a function, reverseList, that takes in the head of a linked list as an argument. The function should reverse the order of the nodes in the linked list in-place and return the new head of the reversed linked list.
*/

/*
INPUT: Node (head)
OUTPUT: Node (reversed linked list head)

RULES 
- it should reverse the order of the nodes in the linked list 
- it should reverse the linked list "in-place"
- it should return the head of the reverse linked list 

Assumptions 
- A valid linked list with at least one Node will always be provided
    - we'll never get NULL as a value 

Mental Model 
The objective is to reverse the order of a singly linked list. To achieve this, we'll manipulate the `next` property of each node to reference the preceding one instead of the succeeding one.

Here's a step-by-step breakdown:
1. Initialize a Variable: Start by initializing a variable, prev, to NULL. This will represent the "previous" node as we traverse the list.
2. Traverse the List: As we move through each node in the linked list:
    - Temporarily store the current node's next reference (so we don't lose access to the remaining part of the list).
    - Update the current node's next property to point to prev (effectively reversing its direction).
    - Move forward in the list by updating the prev to be the current node and the current node to the temporarily stored next node.
3. Reach the End: Continue this process until we've traversed the entire list and the current node points to NULL.
4. Final Step: At this point, prev will be pointing to what used to be the last node, but it's now the first node in our reversed list. Return prev as the new head of the reversed linked list.


DATA STRUCTURE 
- set `prev` to NULL 

while `node` IS NOT NULL keep traversing
    - set `next` to the the current node's next property 
    - set the current node's next property to point to the `prev` node 
    - set `prev` to the current node 
    - set the current node to `next`
- return `prev`

COMPLEXITY 
> N: # of nodes in the linked list 
- Time: O(N)
- Space: O(1)
*/

import Node from '../node.js';

const reverseList = (node) => {
    let prev = null; 

    while (node !== null) {
        // Here, we save the reference to the rest of the linked list
        // otherwise, we'd lose it when we set the `next` property of the current node to the previous node
        const next = node.next;
        
        // Here, we actually execute the reversal. We set the `next` property of the current node to point to `prev` 
        // Like I previously mentioned, if we set this before next = current.next, we'd lose the rest of the linked list reference
        node.next = prev;

        // TRAVERSING THROUGH THE LINKED LIST 
        // We set `prev` to `current` to move the `prev` pointer forward, so now `prev` becomes the current node 
        prev = node;

        // Same as above, we move the current node to the reference stored in `next`, which holds the rest of the linked list
        node = next;
    }
    
    return prev;
};

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;

// a -> b -> c -> d -> e -> f

console.log(JSON.stringify(reverseList(a))); // f -> e -> d -> c -> b -> a


const x = new Node("x");
const y = new Node("y");

x.next = y;

// x -> y

console.log(reverseList(x)); // y -> 


const p = new Node("p");

// p

console.log(reverseList(p)); // p

File: structy/02-linked-list/06-reverse-linked-list/01.py
# reverse list

# Write a function, reverse_list, that takes in the head of a linked list as an argument. The function should reverse the order of the nodes in the linked list in-place and return the new head of the reversed linked list.

"""
INPUT
- head: linked list node

OUTPUT
- reversed linked list 

RULES 
- reverse the linked list "in-place"
    - it mutates the original linked list 
- return the head of the reversed linked list 

Assumptions
- A non-empty linked list will be provided
    - A linked list with at least one node

MENTAL MODEL 
What is a (singly) linked list?
- A linked list is an object that usually has a value and a reference to the next object. 
- The first node of a linked list is known as the "head". 
- The last node of a linked list is known as the "tail", and it has a reference to None. It marks the end of the linked list.

What are we trying to accomplish? 
input: "a" -> "b" -> "c" -> "d" -> None
output: "d" -> "c" -> "b" -> "a" -> None

We need to keep track of 3 key pieces of information (keep references to): 
1. the `previous` node: 
    - previous will be a temporary variable that will store the reverse linked list
2. the `current` node: 
    - this will give us access to the next reference, and flip it
3. the `next` node: 
    - this will store the rest of the linked list 

ALGORITHM 
>> high-level algorithm
1. traverse the linked list 
2. flip the references by pointing "current_node.next" to the previous node
3. return the head of the mutated original linked list

>> mid-level 


"""

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


def reverse_list(head: Node) -> Node:
    prev = None
    current = head

    while current is not None: 
        next_node = current.next    # store the next node 
        current.next = prev         # reverse the pointer of the current node
        # keep advancing the list
        prev = current              # move prev to current 
        current = next_node         # advanced current to next node 

    return prev


a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")
e = Node("e")
f = Node("f")

a.next = b
b.next = c
c.next = d
d.next = e
e.next = f

# a -> b -> c -> d -> e -> f

print(reverse_list(a)) # f -> e -> d -> c -> b -> a -> None


x = Node("x")
y = Node("y")

x.next = y

# x -> y

print(reverse_list(x)) # y -> x -> None
p = Node("p")

# p

print(reverse_list(p)) # p -> None